File: 79-word-search/word-search.java
class Solution {
    public boolean exist(char[][] board, String word) {
        for(int i=0; i<board.length; i++){
            for(int j=0; j<board[0].length; j++){
                if(board[i][j] == word.charAt(0)) {
                    if (dfs(i, j, board, word, 0))
                        return true;
                }
            }
        }
        return false;
    }

    static boolean dfs(int i, int j, char[][] board, String word, int index){
        if(i<0 || j<0 || i>=board.length || j>=board[0].length) return false;
        if(board[i][j]!=word.charAt(index)) return false;
        if(index==word.length()-1) return true;

        char temp = board[i][j];
        board[i][j]='#';
        boolean found = dfs(i+1,j,board,word,index+1) || dfs(i-1,j,board,word,index+1) || dfs(i,j+1,board,word,index+1) || dfs(i,j-1,board,word,index+1);
        board[i][j]=temp;
        return found;
    }
}

File: 79-word-search/word-search.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        rows, cols = len(board), len(board[0])
        
        for r in range(rows):
            for c in range(cols):
                if board[r][c] == word[0] and self._dfs(board, word, r, c, 0, rows, cols, set()):
                    return True
                    
        return False

    def _dfs(self, board: List[List[str]], word: str, r: int, c: int, index: int, rows: int, cols: int, visited: set) -> bool:
        if index == len(word):
            return True
        
        if r < 0 or c < 0 or r >= rows or c >= cols or board[r][c] != word[index] or (r, c) in visited:
            return False
        
        visited.add((r, c))
        
        found = (
            self._dfs(board, word, r + 1, c, index + 1, rows, cols, visited) or
            self._dfs(board, word, r - 1, c, index + 1, rows, cols, visited) or
            self._dfs(board, word, r, c + 1, index + 1, rows, cols, visited) or
            self._dfs(board, word, r, c - 1, index + 1, rows, cols, visited)
        )
        
        visited.remove((r, c))
        return found